_blogposts/2023-04-15-improving-interop.mdx

```diff
@@ -0,0 +1,317 @@
+
+```rescript
+@tag("state")
+type loadingState = | @as("loading") Loading({ready: bool}) | @as("error") Error({message: string}) | @as("done") Done({data: data})
+
+let state = Error({message: "Something went wrong!"})
+```
+
+This will compile to:
+
+```javascript
+var state = {
+  state: "error",
+  message: "Something went wrong!",
+};
+```
+
+Let's break down what we've done to make this work:
+
+- The `@tag` attribute lets you customize the discriminator (default: `TAG`). We're setting that to `"state"` so we map to what the external data looks like.
+- Each variant case has an `@as` attribute. That controls what each variant case is discriminated on (default: the variant case name as string). We're setting all of the cases to their lowercase equivalent, because that's what the external data looks like.
+
+The end result is clean and zero cost bindings to the external data, in a way that previously would require manual runtime conversion.
+
+Now, let's look at a few more real-world examples.
+
+### Binding to TypeScript enums
+
+```typescript
+// direction.ts
+/** Direction of the action. */
+enum Direction {
+  /** The direction is up. */
+  Up = "UP",
+
+  /** The direction is down. */
+  Down = "DOWN",
+
+  /** The direction is left. */
+  Left = "LEFT",
+
+  /** The direction is right. */
+  Right = "RIGHT",
+}
+
+export const myDirection = Direction.Up;
+```
+
+Previously, you'd be forced to use a polymorphic variant for this if you wanted clean, zero-cost interop:
+
+```rescript
+type direction = [#UP | #DOWN | #LEFT | #RIGHT]
+@module("./direction.js") external myDirection: direction = "myDirection"
+```
+
+Notice a few things:
+
+- We're forced to use the names of the enum payload, meaning it won't fully map to what you'd use in TypeScript
+- There's no way to bring over the documentation strings, because polymorphic variants are structural, so there's no one source definition for them to look for docstrings on. This is true _even_ if you annotate with your explicitly written out polymorphic variant definition.
+
+With the new runtime representation, this is how you'd bind to the above enum instead:
+
+```rescript
+/** Direction of the action. */
+type direction =
+  | /** The direction is up. */
+  @as("UP")
+  Up
+
+  | /** The direction is down. */
+  @as("DOWN")
+  Down
+
+  | /** The direction is left. */
+  @as("LEFT")
+  Left
+
+  | /** The direction is right. */
+  @as("RIGHT")
+  Right
+
+@module("./direction.js") external myDirection: direction = "myDirection"
+```
+
+Now, this maps 100% to the TypeScript code, including letting us bring over the documentation strings so we get a nice editor experience.
+
+### String literals
+
+The same logic is easily applied to string literals from TypeScript, only here the benefits is even larger, because string literals have the same limitations in TypeScript that polymorphic variants have in ReScript.
```

Comment:
```suggestion
The same logic is easily applied to string literals from TypeScript, only here the benefit is even larger, because string literals have the same limitations in TypeScript that polymorphic variants have in ReScript.
```